File: kr_stock_trader/strategies.py
"""진입/청산 신호 전략.

국내 현금 계좌는 공매도가 불가하므로 신호는 BUY(신규 매수) / SELL(보유분 청산) / HOLD 세 가지다.
"""

from dataclasses import dataclass
from enum import Enum
from typing import List

from .indicators import ema, rsi, sma
from .models import Candle


class Signal(str, Enum):
    BUY = "BUY"
    SELL = "SELL"
    HOLD = "HOLD"


@dataclass
class StrategyResult:
    signal: Signal
    reason: str


class Strategy:
    name = "base"

    def min_candles(self) -> int:
        raise NotImplementedError

    def evaluate(self, candles: List[Candle]) -> StrategyResult:
        raise NotImplementedError


class MovingAverageCrossoverStrategy(Strategy):
    """단기 EMA 가 장기 EMA 를 상향 돌파하면 매수, 하향 돌파하면 매도 (기본 9/21)."""

    name = "ma_cross"

    def __init__(self, short: int = 9, long: int = 21):
        if short >= long:
            raise ValueError("short 기간은 long 기간보다 작아야 합니다")
        self.short, self.long = short, long

    def min_candles(self) -> int:
        return self.long + 2

    def evaluate(self, candles):
        closes = [c.close for c in candles]
        if len(closes) < self.min_candles():
            return StrategyResult(Signal.HOLD, "캔들 부족")
        s, l = ema(closes, self.short), ema(closes, self.long)
        prev_diff, diff = s[-2] - l[-2], s[-1] - l[-1]
        if prev_diff <= 0 < diff:
            return StrategyResult(Signal.BUY, f"골든크로스 EMA{self.short}/{self.long}")
        if prev_diff >= 0 > diff:
            return StrategyResult(Signal.SELL, f"데드크로스 EMA{self.short}/{self.long}")
        return StrategyResult(Signal.HOLD, f"대기 (EMA{self.short}-EMA{self.long}={diff:,.1f})")


class RsiStrategy(Strategy):
    """RSI 가 과매도선을 아래에서 위로 회복하면 매수, 과매수선을 위에서 아래로 이탈하면 매도."""

    name = "rsi"

    def __init__(self, period: int = 14, oversold: float = 30, overbought: float = 70):
        self.period, self.oversold, self.overbought = period, oversold, overbought

    def min_candles(self) -> int:
        return self.period + 3

    def evaluate(self, candles):
        closes = [c.close for c in candles]
        if len(closes) < self.min_candles():
            return StrategyResult(Signal.HOLD, "캔들 부족")
        r = rsi(closes, self.period)
        prev, cur = r[-2], r[-1]
        if prev < self.oversold <= cur:
            return StrategyResult(Signal.BUY, f"RSI 과매도 탈출 ({prev:.1f}→{cur:.1f})")
        if prev > self.overbought >= cur:
            return StrategyResult(Signal.SELL, f"RSI 과매수 이탈 ({prev:.1f}→{cur:.1f})")
        return StrategyResult(Signal.HOLD, f"대기 (RSI {cur:.1f})")


class CombinedMaRsiStrategy(Strategy):
    """EMA 정배열(단기>장기) 상태에서 RSI 가 과열이 아닐 때만 매수. 역배열 전환 또는 RSI 과열 시 매도."""

    name = "ma_rsi"

    def __init__(self, short: int = 9, long: int = 21, rsi_period: int = 14,
                 rsi_entry_max: float = 60, rsi_exit: float = 75):
        self.ma = MovingAverageCrossoverStrategy(short, long)
        self.rsi_period, self.rsi_entry_max, self.rsi_exit = rsi_period, rsi_entry_max, rsi_exit

    def min_candles(self) -> int:
        return max(self.ma.min_candles(), self.rsi_period + 3)

    def evaluate(self, candles):
        closes = [c.close for c in candles]
        if len(closes) < self.min_candles():
            return StrategyResult(Signal.HOLD, "캔들 부족")
        s, l = ema(closes, self.ma.short), ema(closes, self.ma.long)
        r = rsi(closes, self.rsi_period)[-1]
        uptrend, prev_uptrend = s[-1] > l[-1], s[-2] > l[-2]
        if uptrend and not prev_uptrend and r <= self.rsi_entry_max:
            return StrategyResult(Signal.BUY, f"골든크로스 + RSI {r:.1f}")
        if (prev_uptrend and not uptrend) or r >= self.rsi_exit:
            return StrategyResult(Signal.SELL, f"추세 이탈/과열 (RSI {r:.1f})")
        return StrategyResult(Signal.HOLD, f"대기 (RSI {r:.1f})")


STRATEGIES = {
    MovingAverageCrossoverStrategy.name: MovingAverageCrossoverStrategy,
    RsiStrategy.name: RsiStrategy,
    CombinedMaRsiStrategy.name: CombinedMaRsiStrategy,
}


def build_strategy(name: str, **params) -> Strategy:
    try:
        return STRATEGIES[name](**params)
    except KeyError:
        raise ValueError(f"알 수 없는 전략: {name} (가능: {', '.join(STRATEGIES)})")


def trend_filter_ok(candles: List[Candle], period: int) -> bool:
    """상위 추세 필터: 종가가 장기 이동평균 위에 있을 때만 신규 매수 허용."""
    closes = [c.close for c in candles]
    ma = sma(closes, period)
    if not ma or ma[-1] is None:
        return True  # 데이터가 모자라면 필터를 적용하지 않는다
    return closes[-1] >= ma[-1]

File: kr_stock_app/lib/core/strategies.dart
// 매매 전략. 현금 계좌라 BUY(신규 매수) / SELL(보유분 청산) / HOLD 세 가지 신호만 낸다.
import 'indicators.dart';
import 'models.dart';

enum Signal { buy, sell, hold }

class StrategyResult {
  final Signal signal;
  final String reason;
  const StrategyResult(this.signal, this.reason);
}

abstract class Strategy {
  String get name;
  String get label;
  int get minCandles;
  StrategyResult evaluate(List<Candle> candles);
}

class MaCrossStrategy implements Strategy {
  final int short, long;
  MaCrossStrategy({this.short = 9, this.long = 21}) {
    if (short >= long) throw ArgumentError('short 기간은 long 기간보다 작아야 합니다');
  }
  @override
  String get name => 'ma_cross';
  @override
  String get label => 'MA 크로스 ($short/$long)';
  @override
  int get minCandles => long + 2;

  @override
  StrategyResult evaluate(List<Candle> candles) {
    if (candles.length < minCandles) return const StrategyResult(Signal.hold, '캔들 부족');
    final c = candles.map((e) => e.close).toList();
    final s = ema(c, short), l = ema(c, long);
    final prev = s[s.length - 2]! - l[l.length - 2]!, cur = s.last! - l.last!;
    if (prev <= 0 && cur > 0) return StrategyResult(Signal.buy, '골든크로스 EMA$short/$long');
    if (prev >= 0 && cur < 0) return StrategyResult(Signal.sell, '데드크로스 EMA$short/$long');
    return StrategyResult(Signal.hold, '대기 (EMA 차이 ${cur.toStringAsFixed(1)})');
  }
}

class RsiStrategy implements Strategy {
  final int period;
  final double oversold, overbought;
  RsiStrategy({this.period = 14, this.oversold = 30, this.overbought = 70});
  @override
  String get name => 'rsi';
  @override
  String get label => 'RSI 과매수/과매도';
  @override
  int get minCandles => period + 3;

  @override
  StrategyResult evaluate(List<Candle> candles) {
    if (candles.length < minCandles) return const StrategyResult(Signal.hold, '캔들 부족');
    final r = rsi(candles.map((e) => e.close).toList(), period);
    final prev = r[r.length - 2]!, cur = r.last!;
    if (prev < oversold && cur >= oversold) {
      return StrategyResult(Signal.buy, 'RSI 과매도 탈출 (${prev.toStringAsFixed(1)}→${cur.toStringAsFixed(1)})');
    }
    if (prev > overbought && cur <= overbought) {
      return StrategyResult(Signal.sell, 'RSI 과매수 이탈 (${prev.toStringAsFixed(1)}→${cur.toStringAsFixed(1)})');
    }
    return StrategyResult(Signal.hold, '대기 (RSI ${cur.toStringAsFixed(1)})');
  }
}

class MaRsiStrategy implements Strategy {
  final int short, long, rsiPeriod;
  final double rsiEntryMax, rsiExit;
  MaRsiStrategy({this.short = 9, this.long = 21, this.rsiPeriod = 14, this.rsiEntryMax = 60, this.rsiExit = 75});
  @override
  String get name => 'ma_rsi';
  @override
  String get label => 'MA + RSI 결합';
  @override
  int get minCandles => (long + 2) > (rsiPeriod + 3) ? long + 2 : rsiPeriod + 3;

  @override
  StrategyResult evaluate(List<Candle> candles) {
    if (candles.length < minCandles) return const StrategyResult(Signal.hold, '캔들 부족');
    final c = candles.map((e) => e.close).toList();
    final s = ema(c, short), l = ema(c, long);
    final r = rsi(c, rsiPeriod).last!;
    final up = s.last! > l.last!, prevUp = s[s.length - 2]! > l[l.length - 2]!;
    if (up && !prevUp && r <= rsiEntryMax) {
      return StrategyResult(Signal.buy, '골든크로스 + RSI ${r.toStringAsFixed(1)}');
    }
    if ((prevUp && !up) || r >= rsiExit) {
      return StrategyResult(Signal.sell, '추세 이탈/과열 (RSI ${r.toStringAsFixed(1)})');
    }
    return StrategyResult(Signal.hold, '대기 (RSI ${r.toStringAsFixed(1)})');
  }
}

const strategyNames = {'ma_cross': 'MA 크로스 (9/21)', 'rsi': 'RSI 과매수/과매도', 'ma_rsi': 'MA + RSI 결합'};

Strategy buildStrategy(String name) => switch (name) {
      'ma_cross' => MaCrossStrategy(),
      'rsi' => RsiStrategy(),
      _ => MaRsiStrategy(),
    };

/// 종가가 장기 이동평균 위에 있을 때만 신규 매수 허용. 데이터가 모자라면 통과.
bool trendFilterOk(List<Candle> candles, int period) {
  final m = sma(candles.map((e) => e.close).toList(), period);
  if (m.isEmpty || m.last == null) return true;
  return candles.last.close >= m.last!;
}
